File: mp2i/files/C12/abr.ml
open Hashtbl

let ht = Hashtbl.create 1000;;

type abr = 
 |Vide 
 |Noeud of abr *  int * abr;;
let ninv = ref 0;;

let rec taille ab =
  match ab with
  | Vide -> 0
  | Noeud (sag, _, sad) -> 1 + taille sag + taille sad;;

let rec write_edge a writer =
    match a with
    | Vide -> ()
    | Noeud (sag,e,sad) ->
      match sag, sad with
      | Vide, Vide -> ();
      | Noeud (_, eg, _), Vide ->
          Printf.fprintf writer "%d -> %d\n" e eg;
          Printf.fprintf writer  "I%d [style=invis]\n %d -> I%d [style=invis]\n" !ninv e !ninv;
          ninv := !ninv +1;
          Printf.fprintf writer  "I%d [style=invis]\n %d -> I%d [style=invis]\n" !ninv e !ninv;
          ninv := !ninv +1;
          write_edge sag writer;
      | Vide, Noeud(_,ed,_) ->
          Printf.fprintf writer  "I%d [style=invis]\n %d -> I%d [style=invis]\n" !ninv e !ninv;
        ninv := !ninv +1;
          Printf.fprintf writer  "I%d [style=invis]\n %d -> I%d [style=invis]\n" !ninv e !ninv;
          ninv := !ninv +1;
          Printf.fprintf writer "%d -> %d\n" e ed;
          write_edge sad writer;
      | Noeud (_,eg,_), Noeud (_,ed, _ ) ->
          Printf.fprintf writer "%d -> %d\n" e eg;
          Printf.fprintf writer  "I%d [style=invis]\n %d -> I%d [style=invis]\n" !ninv e !ninv;
          ninv := !ninv +1;
          Printf.fprintf writer "%d -> %d\n" e ed;
          write_edge sag writer;
          write_edge sad writer;
      ;;

let view a =
  let outc = open_out "temp.gv" in
  output_string outc "digraph mygraph {\n";
  write_edge a outc;
  output_string outc "}\n";
  close_out outc;
  ignore (Sys.command "dot -Tpng temp.gv -o temp.png");
  ignore (Sys.command "eog temp.png");;
  

let rec prefixe a =
  match a with
  | Vide -> ()
  | Noeud (g,e,d) -> Printf.printf "%d " e; prefixe g; prefixe d;;


let rec infixe a =
  match a with
  | Vide -> ()
  | Noeud (g,e,d) ->  infixe g; Printf.printf "%d " e; infixe d;;

  let rec postfixe a =
    match a with
    | Vide -> ()
    | Noeud (g,e,d) ->  postfixe g; postfixe d; Printf.printf "%d " e;;

let rec insere abr nv =
  match abr with
  | Vide -> Noeud(Vide,nv,Vide)
  | Noeud(g,v,d) -> if nv<v then Noeud(insere g nv, v, d) else Noeud(g, v, insere d nv);; 

let rec in_abr abr u =
  match abr with
  | Vide -> false
  | Noeud(g,v,d) -> if u=v then true else
                    if u<v then in_abr g u else in_abr d u;;

let () =
  Random.self_init();
  let t = ref Vide in
  let nv = ref (Random.int 100) in
  for i = 0 to 20 do
    while (Hashtbl.mem ht !nv) do
      nv := Random.int 100;
    done;
    Hashtbl.add ht !nv 0;
    t:= insere !t !nv;
  done;
  infixe !t;
  print_newline ();
  view !t;;
  

File: mp2i/files/C12/abr.c
#include <stdio.h>
#include <stdlib.h>

//--8<-- [start:struct]
struct noeud
{
    struct noeud *sag;
    int valeur;
    struct noeud *sad;
};
typedef struct noeud noeud;
typedef noeud *abr;
//--8<-- [end:struct]

void insere(abr *g, int v)
{
    if (*g == NULL)
    {
        noeud *n = malloc(sizeof(noeud));
        n->valeur = v;
        n->sag = NULL;
        n->sad = NULL;
        *g = n;
    }
    else
    {
        if (v < (*g)->valeur)
        {
            insere(&((*g)->sag), v);
        }
        else
        {
            insere(&((*g)->sad), v);
        }
    }
}

int taille(abr a)
{
    if (a == NULL)
    {
        return 0;
    }
    return 1 + taille(a->sag) + taille(a->sad);
    ;
}
void write_edge(FILE *writer, ab g, int *ninv)
{
    if (g->sag != NULL)
    {
        fprintf(writer, "%d", g->valeur);
        fprintf(writer, " -> ");
        fprintf(writer, "%d\n", g->sag->valeur);
        write_edge(writer, g->sag, ninv);
    }
    else
    {
        fprintf(writer, "I%d [style=invis]\n", *ninv);
        fprintf(writer, "%d -> I%d [style=invis]\n", g->valeur, *ninv);
        (*ninv)++;
    }
    fprintf(writer, "I%d [style=invis]\n", *ninv);
    fprintf(writer, "%d -> I%d [style=invis]\n", g->valeur, *ninv);
    (*ninv)++;
    if (g->sad != NULL)
    {
        fprintf(writer, "%d", g->valeur);
        fprintf(writer, " -> ");
        fprintf(writer, "%d\n", g->sad->valeur);
        write_edge(writer, g->sad, ninv);
    }
    else
    {
        fprintf(writer, "I%d [style=invis]\n", *ninv);
        fprintf(writer, "%d -> I%d [style=invis]\n", g->valeur, *ninv);
        (*ninv)++;
    }
}

void view(ab g)
{
    FILE *writer = fopen("temp.gv", "w");
    int n = 0;
    fprintf(writer, "digraph mygraph {\n");
    write_edge(writer, g, &n);
    fprintf(writer, "}\n");
    fclose(writer);
    system("xdot temp.gv &");
}

void prefixe(abr a)
{
    if (a != NULL)
    {
        printf("%d ", a->valeur);
        prefixe(a->sag);
        prefixe(a->sad);
    }
}

int minv(abr *g)
{
    if ((*g)->sag == NULL)
    {
        int t = (*g)->valeur;
        abr td = (*g)->sad;
        free(*g);
        *g=td;
        return t;
    }
    else {return minv (&(*g)->sag);}
}

void supprime(abr *g, int v)
{
    if (*g != NULL)
    {
        if (v < (*g)->valeur)
        {
            supprime(&((*g)->sag), v);
        }
        else if (v > (*g)->valeur)
        {
            supprime(&((*g)->sad), v);
        }
        else
        {
            if ((*g)->sag==NULL && (*g)->sad==NULL)
            {
                free(*g);
                *g = NULL;
            }
            else if ((*g)->sad==NULL) 
            {
                (*g) = (*g)->sag;
                free((*g)->sag);
            }
            else
            {
                int nv =  minv(&(*g)->sad);
                (*g)->valeur = nv;
            }
        }
    }
}

int main()
{
    abr g = NULL;
    insere(&g, 7);
    insere(&g, 5);
    insere(&g, 9);
    insere(&g, 2);
    insere(&g, 11);
    insere(&g, 10);
    insere(&g, 6);
    insere(&g, 8);
    supprime(&g,9);
    supprime(&g,7);
    supprime(&g,2);
    view(g);
    printf("\n");
}